### PR TITLE
(WIP) Implement application decoupling 

### LINKED ISSUE #913: Application decoupling
## Purpose

Allow easier development of multi-platform applications with easy code reuse across them.

Make the entire client-side application and its dependencies, decoupled from the build process and encapsulated inside `src/client`.

## Background

At the moment we have our client application in the `src/client` directory and the build-related tasks in `tools`. Although we have this separation, all the client-side related dependencies and testing configuration are global (`test-main.js`, `karma.config.js`, `protractor.config.js`).

This makes it unpractical to use the seed for development of multiple Angular applications for different platforms which are sharing common features.

For instance, in case we want to develop backend for the current application, which reuses some of the logic which is already defined by services in the client-side, we need to:

- Store all the server-side related dependencies in the `package.json` file where we have the tooling related dependencies and client side related dependencies.
- Have another testing configuration inside `src/server`.

This approach is not modular and has several drawbacks:

- If we need to build only the backend we need to install the dependencies for the backend, client and tooling.
- The configuration is not encapsulated and it could be confusing to have different versions of the same configuration files (for client in the root directory and for the server in `src/server`).

## Proposal

- Move all the files which are related to client-side configuration and logic to `src/client`.
- Have a single build for each individual application (i.e. leave the current build organization unchanged).
- Leave all build related dependencies in the `package.json` in the root of the project and move all others to `src/client/package.json`.

```
.
├── appveyor.yml
├── gulpfile.ts
├── src
│   └── client
│       ├── dist
│       ├── coverage
│       ├── app
│       ├── assets
│       ├── css
│       ├── index.html
│       ├── test-main.js
│       ├── tsconfig.json
│       ├── karma.conf.js
│       ├── package.json
│       ├── protractor.conf.js
│       └── typings.d.ts
├── tools
├── tsconfig.json
├── tslint.json
├── typings
├── package.json
└── typings.json
```

@NathanWalker's advanced seed should benefit a lot from the proposal above.

// cc @ludohenin @d3viant0ne @TheDonDope @nareshbhatia

### PR DESCRIPTION
**Pull request by TheDonDope**
Hey guys, 

i went ahead and started working on implementing the application decoupling as discussed in #913.
The PR is still a a work in progress:

- [X] rename `src/client` to `src/browser` <- IF `browser` is now fine for everyone :D I could change it if you want
- [X] move `karma.conf.js` to `src/browser/karma.conf.js`
- [X] move `test-main.js` to `src/browser/test-main.js`
- [X] move `protractor.conf.js` to `src/browser/protractor.conf.js`
- [X] make sure `$ npm start` is still running fine (update `seed.config.ts`)
- [X] make sure `$ npm test` is still running fine (updating the paths in `karma.start.ts`, `karma.conf.js` and `test-main.js`
- [X] exclude `karma.conf.js`, `protractor.conf.js` and `test-main.js` from the copy process to `dist/**`
- [X] make sure `$ npm run webdriver-start && $ npm run serve.e2e && $ npm run e2e` still runs fine *should work now with the updates of commit https://github.com/mgechev/angular2-seed/pull/926/commits/4b2ca0bb463dcac7bba8cff81d5b7453aa297c8d*
- [ ] implement dedicated `package.json` in `src/browser`
- [ ] update build tasks / configuration to create `dist/browser/dev|prod|test|tmp` directories

Any feedback/help/change requests are welcome as always :)

Have a nice day!

// cc @mgechev @ludohenin @d3viant0ne @NathanWalker @Shyam-Chen

### CODE REVIEW

File: src/browser/karma.conf.js [hunk 1]
```diff
@@ -47,7 +47,7 @@ module.exports = function(config) {
       // suppress annoying 404 warnings for resources, images, etc.
       { pattern: 'dist/dev/assets/**/*', watched: false, included: false, served: true },
 
-      'test-main.js'
+      'src/browser/test-main.js'
```
Comment (TheDonDope) [inline src/browser/karma.conf.js]: Currently `test-main.js` gets copied over to `dist/dev/test-main.js`.
Question: Should it point to the copied `test-main.js` in `dist/dev` or the source `test-main.js` in `src/browser`?

File: tools/tasks/seed/karma.start.ts [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 
+import { APP_SRC } from '../../config';
+
 /**
  * Executes the build process, running all unit tests using `karma`.
  */
 export = (done: any) => {
   new (<any>karma).Server({
-    configFile: join(process.cwd(), 'karma.conf.js'),
+    configFile: join(process.cwd(), APP_SRC, 'karma.conf.js'),
```
Comment (TheDonDope) [inline tools/tasks/seed/karma.start.ts]: Currently `karma.conf.js`gets copied over to `dist/dev`.
Question: Should the `configFile` property point towards the `karma.conf.js` in `dist/dev` or the source `karma.conf.js` in `src/browser`?

File: src/browser/karma.conf.js (hunk 1 above)
Comment (mgechev) [inline src/browser/karma.conf.js]: Should not be copied because it is not used there. We should use the version in `src/browser`.

File: tools/tasks/seed/karma.start.ts (hunk 2 above)
Comment (mgechev) [inline tools/tasks/seed/karma.start.ts]: Same as above.

**Comment by mgechev** [comment]
@TheDonDope awesome job!

File: src/browser/karma.conf.js (hunk 1 above)
Comment (TheDonDope) [inline src/browser/karma.conf.js]: Thanks for the info! I updated the build task accordingly with the commit https://github.com/mgechev/angular2-seed/pull/926/commits/0129daca74c42c9f39da348c25e08069ac579d5e

File: tools/tasks/seed/karma.start.ts (hunk 2 above)
Comment (TheDonDope) [inline tools/tasks/seed/karma.start.ts]: Same as above :) Fixed by commit https://github.com/mgechev/angular2-seed/pull/926/commits/0129daca74c42c9f39da348c25e08069ac579d5e

**Comment by Shyam-Chen** [comment]
@TheDonDope 

Need to change

https://github.com/mgechev/angular2-seed/blob/master/tools/tasks/seed/e2e.ts#L22

**Comment by TheDonDope** [comment]
@Shyam-Chen I already tried that, but it did not have an effect. I think there is another change neccessary in https://github.com/mgechev/angular2-seed/blob/master/package.json#L16

Currently, if you run `$ npm run e2e` protractor gets called directly, but not through the `tools/tasks/seed/e2e.ts`.

**Comment by Shyam-Chen** [comment]
@TheDonDope 

e2e has two parts, one is CI, the other is a local

**Comment by Shyam-Chen** [comment]
see #837

**Comment by TheDonDope** [comment]
@Shyam-Chen yes thank you i found that one too through the history of `package.json`. 
However on current master it is:

```json
    "e2e": "protractor",
```

for the local protractor run.

Shouldnt this command needed to be update to something like:

```json
   "e2e": "gulp build.e2e --color && gulp build.js.e2e --color && gulp e2e --color",
```

**Comment by Shyam-Chen** [comment]
@TheDonDope 
Has been corrected.

https://github.com/mgechev/angular2-seed/blob/master/package.json#L30

**Comment by TheDonDope** [comment]
@Shyam-Chen 
Haha, thanks! But now i am completely confused.
Currently (on current master) in the `package.json` there are 4 different protractor related scripts:

```json
  "scripts": {
    [...]
    "e2e": "protractor",
    "e2e.live": "protractor --elementExplorer",
    [...]
    "e2e.ci": "gulp build.prod --color && gulp build.js.e2e --color && gulp e2e --color",
    "tests.all": "npm test && npm run e2e.ci",
    [...]
  },
```

According to the `README.md`, if you want to execute the e2e tests locally, you have to: 

```bash
# e2e (aka. end-to-end, integration) - In three different shell windows
# Make sure you don't have a global instance of Protractor

# npm run webdriver-update <- You will need to run this the first time
npm run webdriver-start
npm run serve.e2e
npm run e2e

# e2e live mode - Protractor interactive mode
# Instead of last command above, you can use:
npm run e2e.live
```

Is this still correct?

If it is correct, then the gulp script `e2e` would be triggered, which would then execute protractor directly, which then would fail since the `protractor.conf.js` has moved to `src/browser/protractor.conf.js`.

So, since we do have a dedicated gulp task `tools/tasks/seed/e2e.ts` shouldn't be the `"e2e"` script be updated? Or even completely removed and only the `e2e.ci` task be used? (Since it is also included in `tests.all`). 


To sum it up: If i understood everything correctly, my guessed solution to the problem would be the following:

In `package.json`:

```json
  "scripts": {
    [...]
    "e2e": "gulp build.prod --color && gulp build.js.e2e --color && gulp e2e --color",
    [...]
    "tests.all": "npm test && npm run e2e",
    [...]
  },
``` 

- Drop the dedicated `e2e.ci` script and include it into the `e2e` script. (Which is also used in `tests.all` then)

What do you think?

**Comment by Shyam-Chen** [comment]
`e2e.ci` It is a single run for e2e tests.

**Comment by Shyam-Chen** [comment]
rename `e2e.ci` to `e2e.singleRun` (I think)

**Comment by TheDonDope** [comment]
🎉  Alright, from my side i'm finished, looking forward for a code review :)

**Comment by mgechev** [comment]
LGMT. I'll take a better look in the afternoon and merge.

Guys great job!

File: package.json [hunk 3]
```diff
@@ -13,8 +13,9 @@
     "build.test": "gulp build.test --color",
     "build.test.watch": "gulp build.test.watch --color",
     "generate.manifest": "gulp generate.manifest --color",
-    "e2e": "protractor",
-    "e2e.live": "protractor --elementExplorer",
+    "e2e": "protractor src/browser/protractor.conf.js",
+    "e2e.live": "protractor src/browser/protractor.conf.js --elementExplorer",
+    "e2e.singleRun": "gulp build.prod --color && gulp build.js.e2e --color && gulp e2e --color",
     "gulp": "gulp",
     "karma": "karma",
```
Comment (Shyam-Chen) [inline package.json]: Is it going to change?

Comment (TheDonDope) [inline package.json]: Hey @Shyam-Chen, what do you mean?

About `"e2e"` -> I had to add the `src/browser/protractror.conf.js` to it, since it has been moved to `src/browser`. The same for `"e2e.live"`. 

About `"e2e.singleRun"` i renamed `"e2e.cli"` as you suggested to `"e2e.singleRun"` and sorted it alphabetically under `"e2e.live"`(since all other scripts are also in alphabetical order)

**Comment by TheDonDope** [comment]
@mgechev Thanks! :) I just wanted to update the `README.md` to include the new directory structure and discovered that the `coverage` directory now gets generated within `src/browser`. I guess that should not be the case, should it? This must have been a side effect of moving the `*.conf.js` files to `src/browser`. I'll take another look at it to make sure it gets generated in its original place.

**Comment by Shyam-Chen** [comment]
add new task?
```ts
// gulpfile.ts
gulp.task('e2e.singleRun', (done: any) =>
  runSequence('build.prod',
              'build.js.e2e',
              'e2e',
              done));
```

`"e2e.singleRun": "gulp e2e.singleRun --color",`

**Comment by TheDonDope** [comment]
@Shyam-Chen Thanks for the suggestion, i updated it and added this with the commit https://github.com/mgechev/angular2-seed/pull/926/commits/8ad516141744fa85b91e9b01c383bb4a9c3c0218 :)

**Comment by ludohenin** [comment]
Awesome guys.
I think one important bit is missing, `package.json` ?

**Comment by TheDonDope** [comment]
Hey @ludohenin what about the `package.json`? What is missing?

**Comment by TheDonDope** [comment]
@ludohenin Oh sh***, i just had another look at the original proposal of #913, and you are totally right! There was a separate `package.json` proposed inside `src/browser`. Sorry, i completely forgot about that one!

**Comment by ludohenin** [comment]
In @mgechev's proposal, the client platform app comes with its own package.json file.

**Comment by TheDonDope** [comment]
@ludohenin I guess it will not be as easy as to just transfer the browser app specific dependencies to the `src/browser/package.json` and then "it just works", would it? Which build scripts would need an update?

**Comment by TheDonDope** [comment]
Also: Would this mean that in consequence that there would be a separete `node_modules` directory within `src/browser`? And furthermore: Would there need to be another `typings.json` within the `src/browser` directory as well as the corresponding `src/browser/typings` directory?

**Comment by mgechev** [comment]
Hahah, I agree about the :beer:, and about the changes that need to be made.

The build needs to be changed to create `dist/browser/dev|prod|test|tmp`. I think we can live with the current set of typings but `package.json` is something that needs to be moved to `src/browser` because otherwise we need to install all the dependencies for all the applications when having many platforms.

**Comment by TheDonDope** [comment]
:ambulance:  Alright, i hereby call for help on the last necessary changes: :ambulance: 
- Implementing the dedicated `src/browser/package.json 
- Updating the necessary build tasks / configuration

Any snippets / commits / fork branches are welcome :) 

Thanks guys!

**Comment by mgechev** [comment]
I'll put this in my schedule for the next week/during the weekend.

**Comment by TheDonDope** [comment]
Continued in #959 which has this commits pushed to the `issue/913-application-decoupling` branch of this repository.
